src/main/java/org/folio/services/validator/engine/ValidationEngineServiceImpl.java

```diff
@@ -0,0 +1,276 @@
+
+  // Logger
+  private final Logger logger = LoggerFactory
+    .getLogger(ValidationEngineServiceImpl.class);
+  // Timeout to wait for response
+  private int lookupTimeout = Integer
+    .parseInt(MODULE_SPECIFIC_ARGS.getOrDefault("lookup.timeout", "1000"));
+  // Repository component to validation obtain rules
+  private ValidatorRegistryService validatorRegistryProxy;
+  // Http client to call programmatic rules as internal OKAPI endpoints
+  private HttpClient httpClient;
+
+  public ValidationEngineServiceImpl() {
+  }
+
+  public ValidationEngineServiceImpl(final Vertx vertx) {
+    this.validatorRegistryProxy = ValidatorRegistryService
+      .createProxy(vertx, ValidatorHelper.REGISTRY_SERVICE_ADDRESS);
+    initHttpClient(vertx);
+  }
+
+  private void initHttpClient(final Vertx vertx) {
+    HttpClientOptions options = new HttpClientOptions();
+    options.setConnectTimeout(lookupTimeout);
+    options.setIdleTimeout(lookupTimeout);
+    this.httpClient = vertx.createHttpClient(options);
+  }
+
+  /**
+   * Validates received password.
+   * Calls ValidationRegistry service to obtain rules specific to tenant,
+   * runs rules to validate password in one loop,
+   * pushes validation result into result handler to return.
+   *
+   * @param password       received password for validation
+   * @param requestHeaders request headers needed for access backend FOLIO services to perform programmatic rules validation
+   * @param resultHandler  handler with validation results in format <Status, Message>
+   */
+
+  @Override
+  public void validatePassword(final String password,
+                               final Map<String, String> requestHeaders,
+                               final Handler<AsyncResult<JsonObject>> resultHandler) {
+    MultiMap caseInsensitiveHeaders = new CaseInsensitiveHeaders().addAll(requestHeaders);
+    String tenantId = caseInsensitiveHeaders.get(OKAPI_HEADER_TENANT);
+    validatorRegistryProxy.getAllTenantRules(tenantId, 500, 1, "query=state=Enabled", rulesResponse -> {
+      if (rulesResponse.succeeded()) {
+        lookupUser(caseInsensitiveHeaders).setHandler(lookupUserHandler -> {
+          if (lookupUserHandler.succeeded()) {
+            List<Rule> rules = rulesResponse.result().mapTo(RuleCollection.class).getRules();
+            prepareRulesBeforeValidation(rules, lookupUserHandler);
+            Future<List<String>> errorMessagesFuture = validatePasswordByRules(rules, password, caseInsensitiveHeaders);
+            errorMessagesFuture.setHandler(asyncResult -> {
+              if (asyncResult.succeeded()) {
+                prepareResponse(asyncResult.result(), resultHandler);
+              } else {
+                resultHandler.handle(Future.failedFuture(asyncResult.cause().getMessage()));
+              }
+            });
+          }
+        });
+      } else {
+        resultHandler.handle(Future.failedFuture(rulesResponse.cause().getMessage()));
+      }
+    });
+  }
```

Comment:
Consider using guard clauses: https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html
This reduces the identation level and makes the code more easy to understand:
```
    validatorRegistryProxy.getAllTenantRules(tenantId, 500, 1, "query=state=Enabled", rulesResponse -> {
      if (rulesResponse.failed()) {
        resultHandler.handle(Future.failedFuture(rulesResponse.cause()));
        return;
      }
      lookupUser(caseInsensitiveHeaders).setHandler(lookupUserHandler -> {
        if (lookupUserHandler.failed()) {
          resultHandler.handle(Future.failedFuture(lookupUserHandler.cause()));
          return;
        }
        List<Rule> rules = rulesResponse.result().mapTo(RuleCollection.class).getRules();
        prepareRulesBeforeValidation(rules, lookupUserHandler);
        Future<List<String>> errorMessagesFuture = validatePasswordByRules(rules, password, caseInsensitiveHeaders);
        errorMessagesFuture.setHandler(asyncResult -> {
          if (asyncResult.failed()) {
            resultHandler.handle(Future.failedFuture(asyncResult.cause()));
            return;
          }
          prepareResponse(asyncResult.result(), resultHandler);
        });
      });
    });
```

Reply:
Refactored according to the suggestion